src/dag/lazy-store.ts

```diff
@@ -0,0 +1,350 @@
+import {Hash, isTempHash} from '../hash';
+import type * as kv from '../kv/mod';
+import {RWLock} from '../rw-lock';
+import {Chunk, ChunkHasher, createChunk} from './chunk';
+import type {Store, Read, Write} from './store';
+import {getSizeOfValue as defaultGetSizeOfValue} from '../get-size-of-value';
+import type {ReadonlyJSONValue} from '../mod';
+import {GarbargeCollector, HeadChange} from './gc';
+
+export class LazyStore implements Store {
+  private readonly _rwLock = new RWLock();
+  private readonly _heads = new Map<string, Hash>();
+  private readonly _tempChunks = new Map<Hash, Chunk>();
+  private readonly _sourceChunksCache: ChunksCache;
+  private readonly _sourceStore: Store;
+  private readonly _refCounts = new Map<Hash, number>();
```

Comment:
Did you consider having `LazyStore` contain a memstore and thereby share more of the GC code btwn two implementations? Seems like you wouldn't need so many callbacks. What are the downsides of that approach?

Reply:
I think that would be simpler too but it will be less efficient. But lets care about performance later when we know what is slow or not.

Reply:
I explored:
1. having a source dag and a mem dag (for the chunk cache/heads/temp) 
2. using a mem kv store for the cache


Neither really fit because:
1. for both kv store and mem dag the locking is wrong for our needs.  we want to put things into the cache on reads, but we don't want reads to grab a write lock.
2. neither works well for doing lru.  mem dag lacks a delete interface.  neither keep track of access order so we would need to do that on top.  
3. the ref counting semantics are slightly different (due to lazy caching and eviction).  look at how refCounts is manipulated in lazystore ChunkCache.  also retrieving of refs during gc has to know not to go to the source store (otherwise it will load things into cache just for ref counts)

As factored, we get to reuse most of the GC logic.